protocol/deployment-chains.md

```diff
@@ -0,0 +1,261 @@
+- Create multiple L2s with a shared L1 and shared superchain (to test interop without emulating hardfork first).
+
+### Devnet needs
+
+As a devnet user, I need to:
+- Modify the chain template, for experimentation when building new features.
+- Support different "modes" without accidental configuration mixup,
+  e.g. I may to run different hardforks, or run with different DA types.
+
+### Production needs
+
+As a production superchain host, I need to:
+- Perform a new Superchain core deployment (global contracts and implementation-contract deployments).
+- Maintain the Superchain core deployment with upgrades.
+- Not deploy/upgrade an L2 at the same time.
+
+As a production L2 chain deployer, I may need to:
+- Build an L2 on top of a pre-existing superchain core deployment.
+- Upgrade the L2 by attaching to new superchain implementation contracts.
+- Customize just the L2 part of the deployment, while retaining compatibility with the general deployment flow.
+
+Production needs may change; there are many external users and new features. Flexibility is key.
+Making the deployment more modular fits this.
+
+# Alternatives Considered
+
+## Few enshrined deploy-configs
+
+Not all that different from status-quo.
+With copy-paste we can add 1 more chain, and suffix it with `-interop`, to produce a 2-chain test environment.
+For early testing this was drafted in [PR 10777](https://github.com/ethereum-optimism/optimism/pull/10777).
+
+This does not yet unify the common superchain part of the deployment, or any of the artifacts,
+nor supports test-specific changes to parameters.
+
+## Presets and configs
+
+In Eth2 there exist two classes of configuration:
+- "preset" (named, enshrined, compile-time)
+- "config" (customizable, run-time)
+
+This helped reduce complexity by encapsulating and limiting compile-time configuration,
+but also limited severely what could be changed.
+For a state-format it may be good to contain options,
+but for the amount of runtime configuration the OP-Stack has, it seems impractical.
+
+And each variable that remains in the runtime-config still cannot be customized in `op-e2e` tests,
+due to limits in the current `l1-allocs` and `l2-allocs` pre-generated files approach.
+
+## Templating
+
+By separating contract-code and contract-storage, the contract deployments can be reconfigured without recompiling.
+However, this pushes complexity to the user of the deployments data:
+it re-introduces smart-contract deployment concepts into the Go chain-ops.
+E.g. storage-layouts, storage packing, the occasional `immutable` variable, all make customization difficult.
+Not to mention cases where we need entirely different contracts, different hardforks, or more L2 deployments.
+
+The insight here is that we should strongly contain smart-contract configuration to the solidity-scripting side,
+and minimize post-processing in Go.
+
+## Mega test deployment
+
+Instead of generating many separate files per configuration,
```

Comment:
Ah i like this, so basically we would have a single L1 world state with many L2 chains deployed to it, each configured a bit differently (standard, plasma, plasma generic, custom gas token). The deploy script outputs a json file with key/value pairs of contract names -> addresses. That could be the view into a particular system. We could add a new entrypoint to the deploy script that uses `loadAllocs` to read in a previous state, we could also use `CONTRACT_ADDRESSES_PATHS` to populate the implementation addresses in the deploy script so that we don't have to redeploy the same implementations each time

Reply:
Resolving this comment, as we discussed this outside of this PR. In summary it boils down to lots of book-keeping of contract addresses and their configurations, and remapping issues with predeploys on fixed addresses.